src-extras/Database/LSMTree/Extras/MergingRunData.hs

```diff
@@ -0,0 +1,217 @@
+-- similarly to how @ScheduledMergesTest@ does it.
+data MergingRunData t k v b =
+    CompletedMergeData t MR.NumRuns (RunData k v b)
+  | OngoingMergeData t [NonEmptyRunData k v b]  -- ^ at least 2 inputs
+  deriving stock (Show, Eq)
+
+-- | See @mergeInvariant@ in the prototype.
+mergingRunDataInvariant :: MergingRunData t k v b -> Either String ()
+mergingRunDataInvariant = \case
+    CompletedMergeData _ (MR.NumRuns n) _ ->
+      assertI "completed merges are non-trivial (at least two inputs)" $
+        n > 1
+    OngoingMergeData _ rds -> do
+      assertI "inputs to ongoing merges aren't empty" $
+        all nonEmptyRunDataInvariant rds
+      assertI "ongoing merges are non-trivial (at least two inputs)" $
+        length rds > 1
+  where
+    assertI msg False = Left msg
+    assertI _   True  = Right ()
+
+mapMergingRunData ::
+     Ord k'
+  => (k -> k') -> (v -> v') -> (b -> b')
+  -> MergingRunData t k v b -> MergingRunData t k' v' b'
+mapMergingRunData f g h = \case
+    CompletedMergeData t n r ->
+      CompletedMergeData t n $ mapRunData f g h r
+    OngoingMergeData t rs ->
+      OngoingMergeData t $
+        map (NonEmptyRunData . mapRunData f g h . unNonEmptyRunData) rs
+
+type SerialisedMergingRunData t =
+    MergingRunData t SerialisedKey SerialisedValue SerialisedBlob
+
+serialiseMergingRunData ::
+     (SerialiseKey k, SerialiseValue v, SerialiseValue b)
+  => MergingRunData t k v b -> SerialisedMergingRunData t
+serialiseMergingRunData =
+    mapMergingRunData serialiseKey serialiseValue serialiseBlob
+
+{-------------------------------------------------------------------------------
+  QuickCheck
+-------------------------------------------------------------------------------}
+
+labelMergingRunData ::
+     Show t => SerialisedMergingRunData t -> Property -> Property
+labelMergingRunData (CompletedMergeData mt _ rd) =
+      tabulate "merging run state" ["CompletedMerge"]
+    . tabulate "merge type" [show mt]
+    . labelRunData rd
+labelMergingRunData (OngoingMergeData mt rds) =
+      tabulate "merging run state" ["OngoingMerge"]
+    . tabulate "merge type" [show mt]
+    . tabulate "merging run inputs" [showPowersOf 2 (length rds)]
+    . foldr ((.) . labelRunData . unNonEmptyRunData) id rds
+
+instance ( Arbitrary t, Ord k, Arbitrary k, Arbitrary v, Arbitrary b
+         ) => Arbitrary (MergingRunData t k v b) where
+  arbitrary = genMergingRunData arbitrary arbitrary arbitrary
+  shrink = shrinkMergingRunData shrink shrink shrink
+
+genMergingRunData ::
+     forall t k v b. (Arbitrary t, Ord k)
+  => Gen k
+  -> Gen v
+  -> Gen b
+  -> Gen (MergingRunData t k v b)
+genMergingRunData genKey genVal genBlob =
+    QC.oneof
+      [ do
+          mt <- arbitrary
+          numRuns <- MR.NumRuns <$> QC.chooseInt (2, 32)
+          rd <- genRunData genKey genVal genBlob
+          pure (CompletedMergeData mt numRuns rd)
+      , do
+          mt <- arbitrary
+          n  <- QC.chooseInt (2, 8)
```

Comment:
Is it important for tests that there are more than 2 or 3? I feel like we test all the interesting cases with 2 or 3 input runs.

Reply:
Related to https://github.com/IntersectMBO/lsm-tree/pull/592#issuecomment-2687564622

Reply:
I would not make it impossible to have more than 3 runs, it might matter somewhere. But we can make it skew towards smaller values, and on my branch I had just tweaked it to `QC.chooseInt (2, max 2 (s * 8 `div` 100))` to still allow up to 8 runs, but have `57.1% 2 <= n < 4`.  However, it doesn't significantly impact the test runtime, since we `QC.scale (`div` n)` below, so on average we still get the same number of entries.